packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -35,6 +35,81 @@ const int _kPubExitCodeUnavailable = 69;
 
 typedef MessageFilter = String? Function(String message);
 
+/// globalCachePath is the directory in which the content of the localCachePath will be moved in
+void joinCaches({
+  required FileSystem fileSystem,
+  required Directory globalCacheDirectory,
+  required Directory dependencyDirectory,
+}) {
+  for (final FileSystemEntity entity in dependencyDirectory.listSync()) {
+    final String newPath = fileSystem.path.join(globalCacheDirectory.path, entity.basename);
+    if (entity is File) {
+      if (!fileSystem.file(newPath).existsSync()) {
+        entity.copySync(newPath);
+      }
+    } else if (entity is Directory) {
+      if (!globalCacheDirectory.childDirectory(entity.basename).existsSync()) {
+        final Directory newDirectory = globalCacheDirectory.childDirectory(entity.basename);
+        newDirectory.createSync();
+        joinCaches(
+          fileSystem: fileSystem,
+          globalCacheDirectory: newDirectory,
+          dependencyDirectory: entity,
+        );
+      }
+    }
+  }
+}
+
+Directory createDependencyDirectory(Directory pubGlobalDirectory, String dependencyName) {
+  final Directory newDirectory = pubGlobalDirectory.childDirectory(dependencyName);
+  newDirectory.createSync();
+  return newDirectory;
+}
+
+void tryDeleteCache(String dependencyBaseName, Directory globalCachePub, Logger logger) {
+  final Directory dependency = globalCachePub.childDirectory(dependencyBaseName);
+  if (dependency.existsSync()) {
+    try {
+      globalCachePub.childDirectory(dependencyBaseName).deleteSync(recursive: true);
+    }
+    on FileSystemException {
+      throwToolExit('FileSystemException found while deleting a directory, '
+        'try running "flutter pub cache repair"'
+      );
+    }
+  }
+  logger.printWarning('The join of pub-caches failed');
+}
+
+void tryToDeleteLocalCache(Directory localCache, Logger logger) {
```

Comment:
can both this function and `tryDeleteCache` be merged into one function that takes in a `FileSystemEntity`?

Reply:
And then it returns a boolean on whether it succeeded?